### ISSUE TITLE
Submitting usage to collector - Status 409

### ISSUE THREAD

**Issue by rajkiranrbala** [cloudfoundry-incubator/cf-abacus#124]
When I submit usage to collector, I get a http status code 409 even when the payload has new usage entries along with the old entries. Is this the expected behavior.

Payload 1 
```javascript
{
  "usage": [
    {
      "start": 1445637382327,
      "end": 1445637382332,
      "organization_id": "us-south:a3d7fe4d-3cb1-4cc3-a831-ffe98e20cf27",
      "space_id": "aaeae239-f3f8-483c-9dd0-de5d41c38b6a",
      "consumer_id": "app:bbeae239-f3f8-483c-9dd0-de6781c38bab",
      "resource_id": "object-storage",
      "plan_id": "basic",
      "resource_instance_id": "0b39fa70-a65f-4183-bae8-385633ca5c87",
      "measured_usage": [
        {
          "measure": "storage",
          "quantity": 1073741824
        },
        {
          "measure": "light_api_calls",
          "quantity": 1000
        },
        {
          "measure": "heavy_api_calls",
          "quantity": 100
        }
      ]
    },
    {
      "start": 1445637382337,
      "end": 1445637382349,
      "organization_id": "us-south:a3d7fe4d-3cb1-4cc3-a831-ffe98e20cf27",
      "space_id": "aaeae239-f3f8-483c-9dd0-de5d41c38b6a",
      "consumer_id": "app:bbeae239-f3f8-483c-9dd0-de6781c38bab",
      "resource_id": "object-storage",
      "plan_id": "basic",
      "resource_instance_id": "0b39fa70-a65f-4183-bae8-385633ca5c87",
      "measured_usage": [
        {
          "measure": "storage",
          "quantity": 1073741824
        },
        {
          "measure": "light_api_calls",
          "quantity": 1000
        },
        {
          "measure": "heavy_api_calls",
          "quantity": 100
        }
      ]
    }
  ]
}
```
Payload 2

```javascript
{
  "usage": [
    {
      "start": 1445637382337,
      "end": 1445637382349,
      "organization_id": "us-south:a3d7fe4d-3cb1-4cc3-a831-ffe98e20cf27",
      "space_id": "aaeae239-f3f8-483c-9dd0-de5d41c38b6a",
      "consumer_id": "app:bbeae239-f3f8-483c-9dd0-de6781c38bab",
      "resource_id": "object-storage",
      "plan_id": "basic",
      "resource_instance_id": "0b39fa70-a65f-4183-bae8-385633ca5c87",
      "measured_usage": [
        {
          "measure": "storage",
          "quantity": 1073741824
        },
        {
          "measure": "light_api_calls",
          "quantity": 1000
        },
        {
          "measure": "heavy_api_calls",
          "quantity": 100
        }
      ]
    },
    {
      "start": 1445637382352,
      "end": 1445637382363,
      "organization_id": "us-south:a3d7fe4d-3cb1-4cc3-a831-ffe98e20cf27",
      "space_id": "aaeae239-f3f8-483c-9dd0-de5d41c38b6a",
      "consumer_id": "app:bbeae239-f3f8-483c-9dd0-de6781c38bab",
      "resource_id": "object-storage",
      "plan_id": "basic",
      "resource_instance_id": "0b39fa70-a65f-4183-bae8-385633ca5c87",
      "measured_usage": [
        {
          "measure": "storage",
          "quantity": 1073741824
        },
        {
          "measure": "light_api_calls",
          "quantity": 1000
        },
        {
          "measure": "heavy_api_calls",
          "quantity": 100
        }
      ]
    }
  ]
}
```

**Comment by jsdelfino** [comment]
Yes, the request will be rejected with a 409 if you have a duplicate in there (duplicate being defined as same resource id, start and end time). Looks like 2nd usage of your 1st request is a duplicate of the 1st usage of your 2nd request.

There was a bug in the duplicate detection code before where duplicates were not always correctly detected. This bug has been fixed this week, and that's the reason for getting a 409.

**Comment by piotrprzybylski** [comment]
So if the submission contains even one duplicate, the whole set is rejected? or the new records are accepted, but the system returns 409 to indicate a duplicate ?

**Comment by jsdelfino** [comment]
Currently the whole set gets rejected, as we need a clear indication of the decision on the batch: either 201 the batch is accepted and you get back a URL where you can retrieve it later, or 409 indicating that the batch is rejected. Service providers can also post individual usage docs (batches of 1 usage doc) if they'd like to get a 201 vs 409 decision on each individual usage doc.

**Comment by piotrprzybylski** [comment]
This can prevent 'incremental submissions' in case of failure on submitter side.  
Consider the following situation - provider submits 5 records.  After getting 201, it tries to log success, but that step fails. While provider pauses, the new usage is generated. If the new record is added to the next (re)submission batch, it will not be accepted as the whole batch is rejected.

**Comment by jsdelfino** [comment]
Supporting that would require turning the duplicate detection off in the collector while keeping it in the other services down the pipeline, starting with the metering service. This might not be a bad idea although we'll still need to find a way to notify the service providers about these duplicates after they've been accepted by the collector. I'm going to investigate to understand the work involved.

**Comment by piotrprzybylski** [comment]
Alternatively, the collector could keep detecting duplicates but accept and pass downstream only new submissions. Same issue that you mention - notifying submitter about duplicates present in the submission would still need to be addressed.

**Comment by hsiliev** [comment]
Accepting only new submissions would be extremely helpful for the cf-bridge as well. TL;DR: we have the same problem as the one described by @piotrprzybylski

Currently we batch the requests The entries in the batch depend on the events on the CC side which does not seem to be with guaranteed order for a 1 minute. We therefore simply ignore events newer that 1 minute and will resubmit the usage hoping the DB will be order in a while.

Another issue is when the bridge restarts - it does not have persistent storage and simply starts from the begining. If we want to keep it simple and stupid (KISS) we will need a way to inform abacus to ignore the duplicates.

I raised an issue #128 to track the changes in cf-bridge (if needed).

**Comment by hsiliev** [comment]
In case of 409 I'm submitting usage one by one until I got 201. Then I activate batching. This seems to solve the problem.

**Comment by jsdelfino** [comment]
This issue was fixed in commits 635c52f (turn off duplicate detection in usage collector) and da7e29a (produce unique doc ids in usage collector).

Commit da7e29a incorrectly referenced issue #132 but it should have referenced this issue instead.

Commit 635c52f also mentioned the idea of storing the duplicate docs in an 'errors' db but that's not needed at this point to fix this particular issue, as the duplicate docs are stored in the usage collector's db anyway so they're not lost even if they're detected as duplicates and skipped by the meter service later.
